### PR TITLE
feat: create macro to generate TryFrom trait implementation for events

### PR DESCRIPTION
**Pull request by cgorenflo**

### CODE REVIEW

File: packages/events_derive/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        use error_stack::{IntoReport as _, ResultExt as _};
+
+        impl TryFrom<&events::Event> for #event_struct {
+            type Error = error_stack::Report<events::Error>;
+
+            fn try_from(event: &events::Event) -> core::result::Result<Self, Self::Error> {
+                match event {
+                    events::Event::Abci { event_type, attributes } if event_type.as_str() == #event_type => {
+                        let event =
+                            TestEvent::deserialize(serde::de::value::MapDeserializer::new(attributes.clone().into_iter()))
```
Comment (sammy1991106) [inline packages/events_derive/src/lib.rs:24]:
```suggestion
                            #event_struct::deserialize(serde::de::value::MapDeserializer::new(attributes.clone().into_iter()))
```

File: packages/events_derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                                    #event_struct_name.to_string()),
+                                )?;
+                        Ok(event)
+                    }
+                    _ => Err(events::Error::EventTypeMismatch(#event_type.to_string())).into_report(),
+                }
+            }
+        }
+
+        impl TryFrom<events::Event> for TestEvent {
```
Comment (sammy1991106) [inline packages/events_derive/src/lib.rs:37]:
```suggestion
        impl TryFrom<events::Event> for #event_struct {
```

File: packages/events_derive/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        #input
+
+        use error_stack::{IntoReport as _, ResultExt as _};
+
+        impl TryFrom<&events::Event> for #event_struct {
+            type Error = error_stack::Report<events::Error>;
+
+            fn try_from(event: &events::Event) -> core::result::Result<Self, Self::Error> {
+                match event {
+                    events::Event::Abci { event_type, attributes } if event_type.as_str() == #event_type => {
```
Comment (sammy1991106) [inline packages/events_derive/src/lib.rs:22]:
```suggestion
                    events::Event::Abci { event_type, attributes } if event_type == #event_type => {
```

File: packages/events_derive/tests/macro.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
+use error_stack::Result;
+use serde::Deserialize;
+
+#[allow(dead_code)]
```
Comment (sammy1991106) [inline packages/events_derive/tests/macro.rs:4]: Why?

File: packages/events_derive/tests/macro.rs
Comment (sammy1991106) [inline packages/events_derive/tests/macro.rs:1]: Can't these tests sit in a `test` mod in lib.rs?

File: packages/events_derive/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
+use proc_macro::TokenStream;
+use quote::quote;
+
+#[proc_macro_attribute]
+pub fn tryfrom(arg: TokenStream, input: TokenStream) -> TokenStream {
```
[the remark this answers is not in the document]
Comment (sammy1991106) [reply packages/events_derive/src/lib.rs:5]: Can you add some comments for this macro?

**Review by sammy1991106 — CHANGES_REQUESTED** [review]

File: packages/events_derive/tests/macro.rs (hunk 4 above)
Comment (cgorenflo) [reply packages/events_derive/tests/macro.rs:4]: clippy ignores debug printing for the usage analysis, so both fields are flagged as not used

File: packages/events_derive/tests/macro.rs
Comment (cgorenflo) [reply packages/events_derive/tests/macro.rs:1]: it's not allowed

File: packages/events/src/event.rs [hunk 6]
```diff
@@ -9,3 +9,13 @@ pub enum Event {
         attributes: serde_json::Map<String, serde_json::Value>,
     },
 }
+
+impl Event {
+    pub fn block_begin(height: impl Into<block::Height>) -> Self {
```
Comment (sammy1991106) [inline packages/events/src/event.rs:14]: 👍

**Review by sammy1991106 — APPROVED** [review]

**Review by haiyizxx — APPROVED** [review]

**Pull request merged**